internal/argocd/health.go

```diff
@@ -217,58 +224,123 @@ func (h *applicationHealth) GetApplicationHealth(
+					if client.IgnoreNotFound(err) == nil {
+						err = fmt.Errorf("unable to find Argo CD Application %q in namespace %q", key.Name, key.Namespace)
+					}
+					return kargoapi.HealthStateUnknown, healthStatus, syncStatus, err
 				}
-				return kargoapi.HealthStateUnknown, healthStatus, syncStatus, err
 			}
 		}
 	}
 
 	// With all the above checks passed, we can now assume the Argo CD
 	// Application's health state is reliable.
-	healthState, err := stageHealthForAppHealth(app)
+	healthState, err := stageHealthForAppHealth(ctx, app)
 	return healthState, healthStatus, syncStatus, err
 }
 
 // stageHealthForAppSync returns the v1alpha1.HealthState for an Argo CD
 // Application based on its sync status.
-func stageHealthForAppSync(app *argocd.Application, revision string) (kargoapi.HealthState, error) {
+func stageHealthForAppSync(
+	ctx context.Context,
+	app *argocd.Application,
+	revisions []string) (kargoapi.HealthState, error) {
+	logger := logging.LoggerFromContext(ctx).WithValues("appName", app.GetName(), "revisions", revisions)
+	logger.Debug("About to determine stage health based on app sync status.")
 	switch {
-	case revision == "":
+	case revisions == nil || (len(revisions) == 1 && revisions[0] == ""):
+		logger.Debug("Desired revision not set, assuming healthy.")
 		return kargoapi.HealthStateHealthy, nil
 	case app.Operation != nil && app.Operation.Sync != nil,
 		app.Status.OperationState == nil || app.Status.OperationState.FinishedAt.IsZero():
+		logger.Debug("Application in sync operation, assuming Unknown.")
 		err := fmt.Errorf(
 			"Argo CD Application %q in namespace %q is being synced",
 			app.GetName(),
 			app.GetNamespace(),
 		)
 		return kargoapi.HealthStateUnknown, err
-	case app.Status.Sync.Revision != revision:
-		err := fmt.Errorf(
-			"Argo CD Application %q in namespace %q is out of sync",
-			app.GetName(),
-			app.GetNamespace(),
-		)
-		return kargoapi.HealthStateUnhealthy, err
+
 	default:
+		// Apps may have multiple revisions in the list of revisions, so we need to check in both places.
+
+		// Trivial case where app has only a single source and revision is set.
+		singleSourceRevision := app.Status.Sync.Revision
+		if !app.IsMultisource() {
+			if len(revisions) == 1 && revisions[0] == singleSourceRevision {
+				return kargoapi.HealthStateHealthy, nil
+			}
+
+			msg := fmt.Sprintf(
+				"Desired revision %q does not match current revision %q of Application %q "+
+					"in namespace %q, assuming unhealthy.",
+				revisions,
+				singleSourceRevision,
+				app.GetName(),
+				app.GetNamespace(),
+			)
+
+			return kargoapi.HealthStateUnhealthy, errors.New(msg)
+		}
+
+		multiSourceRevisions := app.Status.Sync.Revisions
+		// Apps with multiple sources pointed at the same Git repository can only have the same revision
+		// for all sources because ArgoCD does not support the alternative.
+		// We follow ArgoCD shadow-array implementation here that preserves the order of app.spec.sources for
```

Comment:
Shadow array isn't a well-defined or well-understood term. I think it would make it easier to understand if we were more explicit that the list of revisions is indexed the same as the list of sources. "Parallel arrays" also would be well-understood nomenclature.

Reply:
Applies in two spots, I think.